From 69d2a26c4313acd56914cccca1d864f5ef150bf9 Mon Sep 17 00:00:00 2001
From: Akira Ajisaka <aajisaka@apache.org>
Date: Fri, 18 Dec 2015 13:12:33 +0900
Subject: [PATCH 1058/2748] HADOOP-12588. Fix intermittent test failure of
 TestGangliaMetrics. Contributed by Masatake
 Iwasaki.

(cherry picked from commit bd5e207432c54cc75fd8c040e232241678987e53)

Change-Id: I9d2bfd5c3eb5b7d59ea011756ef2de54877e5194
---
 .../hadoop/metrics2/impl/TestGangliaMetrics.java   |   10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/metrics2/impl/TestGangliaMetrics.java b/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/metrics2/impl/TestGangliaMetrics.java
index e097a0f..c386e36 100644
--- a/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/metrics2/impl/TestGangliaMetrics.java
+++ b/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/metrics2/impl/TestGangliaMetrics.java
@@ -29,7 +29,9 @@
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
+import java.util.concurrent.CopyOnWriteArrayList;
 
+import org.apache.commons.io.Charsets;
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
 import org.apache.hadoop.metrics2.AbstractMetric;
@@ -145,7 +147,7 @@ public void testTagsForPrefix() throws Exception {
   private void checkMetrics(List<byte[]> bytearrlist, int expectedCount) {
     boolean[] foundMetrics = new boolean[expectedMetrics.length];
     for (byte[] bytes : bytearrlist) {
-      String binaryStr = new String(bytes);
+      String binaryStr = new String(bytes, Charsets.UTF_8);
       for (int index = 0; index < expectedMetrics.length; index++) {
         if (binaryStr.indexOf(expectedMetrics[index]) >= 0) {
           foundMetrics[index] = true;
@@ -188,13 +190,13 @@ private void checkMetrics(List<byte[]> bytearrlist, int expectedCount) {
    * hence all the captured byte arrays were pointing to one instance.
    */
   private class MockDatagramSocket extends DatagramSocket {
-    private ArrayList<byte[]> capture;
+    private List<byte[]> capture;
 
     /**
      * @throws SocketException
      */
     public MockDatagramSocket() throws SocketException {
-      capture = new  ArrayList<byte[]>();
+      capture = new CopyOnWriteArrayList<byte[]>();
     }
 
     /* (non-Javadoc)
@@ -211,7 +213,7 @@ public void send(DatagramPacket p) throws IOException {
     /**
      * @return the captured byte arrays
      */
-    ArrayList<byte[]> getCapturedSend() {
+    List<byte[]> getCapturedSend() {
       return capture;
     }
   }
-- 
1.7.9.5

